src/Common/src/System/Net/Internals/SocketExceptionFactory.cs

```diff
@@ -6,12 +6,41 @@
 
 namespace System.Net.Internals
 {
-    internal class SocketExceptionFactory
+    internal static partial class SocketExceptionFactory
     {
+        private class ExtendedSocketException : SocketException
+        {
+            private readonly EndPoint _endPoint;
+
+            public ExtendedSocketException(int errorCode, EndPoint endPoint)
+                : base(errorCode)
+            {
+                _endPoint = endPoint;
+            }
+
+            public ExtendedSocketException(SocketError socketError, int platformError)
+                : base((int)socketError)
+            {
+                HResult = platformError;
+            }
+
+            public override string Message
+            {
+                get
+                {
+                    if (_endPoint != null)
+                    {
+                        return base.Message + " " + _endPoint.ToString();
+                    }
+
+                    return base.Message;
+                }
+            }
```

Comment:
Nit: This could be simplified, e.g.:

```csharp
public override string Message
{
    get
    {
        return _endPoint != null ?
            base.Message + " " + _endPoint.ToString() :
            base.Message;
    }
}
```

Or even more with an expression-bodied member, e.g.:

```csharp
public override string Message => _endPoint != null ?
    base.Message + " " + _endPoint.ToString() :
    base.Message;
```